designs/2024-deprecated-rule-metadata/README.md

```diff
@@ -0,0 +1,268 @@
+
+However, this update could be covered in a blog post about general rule documentation best practices, if anyone ever has an interest in writing something like that.
+
+## Drawbacks
+
+<!--
+    Why should we *not* do this? Consider why adding this into ESLint
+    might not benefit the project or the community. Attempt to think
+    about any opposing viewpoints that reviewers might bring up.
+
+    Any change has potential downsides, including increased maintenance
+    burden, incompatibility with other tools, breaking existing user
+    experience, etc. Try to identify as many potential problems with
+    implementing this RFC as possible.
+-->
+
+There are some limited [backwards compatibility](#backwards-compatibility-analysis) concerns for third-party tooling.
+
+## Backwards Compatibility Analysis
+
+<!--
+    How does this change affect existing ESLint users? Will any behavior
+    change for them? If so, how are you going to minimize the disruption
+    to existing users?
+-->
+
+Existing rules will continue to be backwards-compatible with the new format.
+
+Changing the format of these properties mainly affects third-party documentation tooling and websites that use this information, and not ESLint users nor ESLint plugins directly.
+
+For the most part, the new `meta.deprecated` format should be backwards-compatible, as code is often written to check simply for a truthy value inside of `meta.deprecated`, e.g. `if (rule.meta.deprecated) { /* ... */ }`, which will continue to work as expected. If code checks specifically for the boolean `true` value in `meta.deprecated`, or retrieves rule names from `meta.replacedBy`, then it would need to be updated.
+
+Overall, a limited number of third-party tools that might be affected, and these should be trivial to fix when impacts are discovered.
+
+We do not need to consider this to be a breaking change in terms of [ESLint's semantic versioning policy](https://github.com/eslint/eslint#semantic-versioning-policy).
+
+## Alternatives
+
+<!--
+    What other designs did you consider? Why did you decide against those?
+
+    This section should also include prior art, such as whether similar
+    projects have already implemented a similar feature.
+-->
+
+### Do nothing
+
+This would leave the current `meta.deprecated` and `meta.replacedBy` properties as they are, which would continue to be ambiguous and limited in the information they can provide.
+
+### Consolidate `meta.replacedBy` into `meta.deprecated.replacedBy`
+
+This alternative was strongly considered. If we were to design things from scratch, we might choose this approach.
+
+Pros:
+
+- Organizational tidiness
+- Avoids the inconsistent situation where `meta.deprecated` is `false` or omitted but `meta.replacedBy` has a value, resulting in ambiguity about whether the rule is deprecated or not
```

Comment:
This is an open design question. If we stick with two separate properties, what does it mean for `meta.deprecated` to be `false` or omitted but `meta.replacedBy` to contain a value? It seems implied that a rule that has replacements would be deprecated, but it might not be explicitly marked as so which is concerning.

Reply:
Perhaps this is an opportunity for a `RuleTester` check?

Reply:
Thinking from the perspective of types: this is tough to represent precisely in TypeScript. Either...
* ... the types are straightforward (`{ deprecated?: boolean; replacedBy?: ReplaceBy }`) and allow not-good situations (`{ replacedBy: [...] }`)
* ... the types are something more complex such as a union of objects, and are more difficult to understand

Complex types are generally a symptom of complex code ideas. Which I think is true here. It's not intuitive to answer these questions.

Not saying that TypeScript specifically is a reason to do anything 😄 - just that its type system is a good way to sniff out conception complexity.